iterators task (2)/student/include/impl.h

```diff
@@ -1,33 +1,53 @@
 #pragma once
+
 #include <map>
 #include <iterator>
 #include <type_traits>
+#include <utility>
 
 template <class Iterator, bool IsFirst, class T = typename std::iterator_traits<Iterator>::value_type>
 struct pair_iterator : public Iterator
 {
 };
 
-/**
- * @todo Implement iterator for pair key type
- */
 template<class Iterator, class T>
 struct pair_iterator<Iterator, true, T> : public Iterator
 {
-    /** @todo Iterator traits */
-    /** @todo Iterator operations */
-    /** @todo Constructor from original iterator */
+    using value_type = typename T::first_type;
+    using reference =
+        typename std::conditional<
+            std::is_const<typename std::remove_reference<typename std::iterator_traits<Iterator>::reference>::type>::value,
+            const value_type&,
+            value_type&
+        >::type;
+    using pointer = typename std::add_pointer<reference>::type;
+
+    explicit pair_iterator(Iterator iterator) : Iterator(std::move(iterator)) {}
+
+    reference operator*()
+    {
+        return (*this)->first;
+    }
 };
 
-/**
- * @todo Implement iterator for pair value type
- */
 template<class Iterator, class T>
 struct pair_iterator<Iterator, false, T> : public Iterator
 {
-    /** @todo Iterator traits */
-    /** @todo Iterator operations */
-    /** @todo Constructor from original iterator */
+    using value_type = typename T::second_type;
+    using reference =
+        typename std::conditional<
+            std::is_const<typename std::remove_reference<typename std::iterator_traits<Iterator>::reference>::type>::value,
```

Comment:
Nice!
Consider is_const trait extraction to a named struct

Reply:
Done! But the merge request has been closed... Anyway it was interesting!